tests/realtikvtest/txntest/txn_test.go

```diff
@@ -532,3 +534,66 @@ func TestCheckTxnStatusOnOptimisticTxnBreakConsistency(t *testing.T) {
 	tk2.MustExec("admin check table t2")
 	tk2.MustQuery("select * from t2 order by id").Check(testkit.Rows("1 10", "2 11"))
 }
+
+func TestDMLWithAddForeignKey(t *testing.T) {
+	defer config.RestoreFunc()()
+	config.UpdateGlobal(func(conf *config.Config) {
+		conf.TiKVClient.AsyncCommit.SafeWindow = 10 * time.Second
+		conf.TiKVClient.AsyncCommit.AllowedClockDrift = 500 * time.Millisecond
+	})
+
+	store := realtikvtest.CreateMockStoreAndSetup(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("set global tidb_enable_1pc='OFF';")
+	tk.MustExec("set global tidb_enable_metadata_lock='OFF';")
+	tk.MustExec("set global tidb_enable_async_commit='ON'")
+
+	tkDML := testkit.NewTestKit(t, store)
+	tkDML.MustExec("use test")
+
+	tkDDL := testkit.NewTestKit(t, store)
+	tkDDL.MustExec("use test")
+	tkDDL.MustExec("create table parent (id int primary key, val int, index(val));")
+	tkDDL.MustExec("create table child (id int primary key, val int, index(val));")
+
+	// The fail path of this test is:
+	// tk:     INSERT -> ... -> Wait                                       -> PreWrite -> ... -> Async Commit -> Wait    -> ... -> Success.
+	// tkDDL:            DDL -> StateWriteOnly -> checkForeignKeyConstrain -> DDL                             -> Success
+	// After fixing, either the `tkDDL` or `tk` will fail.
+	failpoint.Enable("tikvclient/beforePrewrite", "pause")
+	defer failpoint.Disable("tikvclient/beforePrewrite")
+	failpoint.EnableCall("github.com/pingcap/tidb/pkg/ddl/afterCheckForeignKeyConstrain", func() {
+		require.NoError(t, failpoint.Disable("tikvclient/beforePrewrite"))
+	})
```

Comment:
If you won't cherry-pick this to old versions, maybe use `testfailpoint.Enable` is more concise.